### ISSUE TITLE
Reproduce the commense results on Boolq

### ISSUE THREAD

**Issue by Zhenyu001225** [AGI-Edgerunners/LLM-Adapters#64]
When I'm doing the evaluation, should I use _--load_8bit_? I'm trying to reproduce the results of LLaMa-7B-LoRA

Finetune:
`CUDA_VISIBLE_DEVICES=8 python finetune.py   --base_model 'yahma/llama-7b-hf'   --data_path './ft-training_set/commonsense_170k.json'   --output_dir './trained_models/llama-7b-lora-commonsense/'   --batch_size 16  --micro_batch_size 4   --num_epochs 3   --learning_rate 3e-4   --cutoff_len 256   --val_set_size 120 --eval_step 80 --save_step 80  --adapter_name lora --target_modules '["q_proj", "k_proj", "v_proj", "up_proj", "down_proj"]' --lora_r 32 --lora_alpha 64`

Evaluate:
CUDA_VISIBLE_DEVICES=3 python commonsense_evaluate.py \
    --model LLaMA-7B \
    --adapter LoRA \
    --dataset  boolq \
    --batch_size 1 \
    --base_model 'yahma/llama-7b-hf' \
    --lora_weights './trained_models/llama-7b-lora-commonsense/'

But the result is only 57.5 compared with the table 68.9..
Could you provide me with some insights here?

**Comment by Zhenyu001225** [comment]
And for PIQA the result is 74.6 compared with data in table 80.7.
For Siqa the result is 60.8 compared with data in table 77.4
Should I finetune again? Or adjusting any of the hypermeters

**Comment by lucasliunju** [comment]
Hi May I ask whether you solve this issue now?

**Comment by wutaiqiang** [comment]
btw, I find that a larger batch size would lead to some bad output while bsz=1 not.

**Comment by lucasliunju** [comment]
@wutaiqiang Yes, I also find this problem and bsz=1 can solve the most case, it can still output BAD result for some case.

**Comment by wutaiqiang** [comment]
In my case, the results are even better than reported. You should use one GPU in finetuning.

**Comment by wutaiqiang** [comment]
69.44 | 80.79 | 79.32 | 84.2 | 81.61 | 80.34 | 64.93 | 76.8

boolq | piqa | social_i_qa | hellaswag | winogrande | ARC-Easy | ARC-Challenge | openbookqa

 
</body>
</html>

**Comment by wutaiqiang** [comment]
For llama 7B + lora

**Comment by lucasliunju** [comment]
Hi @wutaiqiang Thanks for your data point. I try to change the base model from "float16" to "float32" or "bfloat16" and I find the output result is not very stable.

**Comment by Zhenyu001225** [comment]
> Hi May I ask whether you solve this issue now?

Hi, I change the version of transformers to 4.35.0 and when doing evaluation batch_size=1.

Now the results are :

<!DOCTYPE html>
Model | Gsm8k | SVAMP | AuQA | MultiArith | SingleEq | AddSub
-- | -- | -- | -- | -- | -- | --
LLama-7B-LoRA-math | 37.9 | 47.0 | 19.68 | 97.5 | 85.83 | 83.54


<!DOCTYPE html>
Model | BoolQ | SiQA | SIQA | Hellaswag | Winogrande | ARC-c | ARC-e | OpenBookQA | Average
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
LLama-7B-LoRA-Commense | 64.01 | 80.25 | 77.28 | 76.50 | 79.79 | 62.54 | 77.31 | 77.4 | 74.39

**Comment by Zhenyu001225** [comment]
> For llama 7B + lora

Hi, what is the version of transformers in your case?

**Comment by wutaiqiang** [comment]
4.32.1

**Comment by Zhenyu001225** [comment]
> 4.32.1

Thank you so much~ I'll try again

**Comment by clarenceluo78** [comment]
> 69.44 | 80.79 | 79.32 | 84.2 | 81.61 | 80.34 | 64.93 | 76.8
> 
> boolq | piqa | social_i_qa | hellaswag | winogrande | ARC-Easy | ARC-Challenge | openbookqa

Hi there, I want to ask if you use the 8bit quantiztation when reproduce?

**Comment by Zhenyu001225** [comment]
> > 69.44 | 80.79 | 79.32 | 84.2 | 81.61 | 80.34 | 64.93 | 76.8
> > boolq | piqa | social_i_qa | hellaswag | winogrande | ARC-Easy | ARC-Challenge | openbookqa
> 
> Hi there, I want to ask if you use the 8bit quantiztation when reproduce?

I didn't open the 8-bit quantization

**Comment by wutaiqiang** [comment]
After rerun, the results are


68.13 | 80.3 | 78.45 | 83.11 | 80.66 | 77.23 | 65.78 | 79.4
-- | -- | -- | -- | -- | -- | -- | --
boolq | piqa | social_i_qa | hellaswag | winogrande | ARC-Easy | ARC-Challenge | openbookqa

**Comment by AaronZLT** [comment]
@wutaiqiang have you tried the math finetuning?

**Comment by wutaiqiang** [comment]
Not yet @AaronZLT

**Comment by wutaiqiang** [comment]
btw, I find the results quite unstable, try for more times and you would get quite different results

**Comment by AaronZLT** [comment]
Hi, @lucasliunju @Zhenyu001225 

> @wutaiqiang Yes, I also find this problem and bsz=1 can solve the most case, it can still output BAD result for some case.

Does the 'bsz=1' here means the batchsize in finetuning or evaluation? Since in general I think the batchsize in evaluation won't impact the eval result, otherwise it will be very weird. The evaluation result from this repo is quite different from the lm-eval-harness (which is the official eval repo used by huggingface-open-llm-leaderboard), the result of lm-eval is poor.

Btw, after finetuning on the dataset commonsense-170k, the performance on 5-shot mmlu drops, which is worse then the base model.

**Comment by wutaiqiang** [comment]
Yes, bsz means the batch size.

you ca refer to:
https://openreview.net/pdf?id=9MDjKb9lGi

@AaronZLT

**Comment by wutaiqiang** [comment]
Also this one:

https://www.reddit.com/r/LocalLLaMA/comments/19dn2to/inconsistencies_in_llm_outputs_single_vs_batched/

**Comment by Leopold1423** [comment]
when bsz>1, the attention_mask should also be used in the generation: 
```prompts = [generate_prompt(instruction, input) for instruction in instructions]
        inputs = tokenizer(prompts, return_tensors="pt", padding=True)
        input_ids = inputs["input_ids"].to(device)
        generation_config = GenerationConfig(
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            num_beams=num_beams,
            **kwargs,
        )
        with torch.no_grad():
            generation_output = model.generate(
                input_ids=input_ids,
                attention_mask=inputs['attention_mask'].to(device),   # add attention mask here 
                generation_config=generation_config,
                return_dict_in_generate=True,
                output_scores=True,
                max_new_tokens=max_new_tokens,
            )
        s = generation_output.sequences
        outputs = tokenizer.batch_decode(s, skip_special_tokens=True)

**Comment by Leopold1423** [comment]
> when bsz>1, the attention_mask should also be used in the generation:
> 
> ```
>         inputs = tokenizer(prompts, return_tensors="pt", padding=True)
>         input_ids = inputs["input_ids"].to(device)
>         generation_config = GenerationConfig(
>             temperature=temperature,
>             top_p=top_p,
>             top_k=top_k,
>             num_beams=num_beams,
>             **kwargs,
>         )
>         with torch.no_grad():
>             generation_output = model.generate(
>                 input_ids=input_ids,
>                 attention_mask=inputs['attention_mask'].to(device),   # add attention mask here 
>                 generation_config=generation_config,
>                 return_dict_in_generate=True,
>                 output_scores=True,
>                 max_new_tokens=max_new_tokens,
>             )
>         s = generation_output.sequences
>         outputs = tokenizer.batch_decode(s, skip_special_tokens=True)
> ```

by the way, in evaluate.py, when using bsz>1, use_cache=False shall be removed or changed to True; otherwise, the ouput will be a mess, can anyone explain why?

**Comment by yinhengw** [comment]
> > Hi May I ask whether you solve this issue now?
> 
> Hi, I change the version of transformers to 4.35.0 and when doing evaluation batch_size=1.
> 
> Now the results are :
> 
> Model	Gsm8k	SVAMP	AuQA	MultiArith	SingleEq	AddSub
> LLama-7B-LoRA-math	37.9	47.0	19.68	97.5	85.83	83.54
> Model	BoolQ	SiQA	SIQA	Hellaswag	Winogrande	ARC-c	ARC-e	OpenBookQA	Average
> LLama-7B-LoRA-Commense	64.01	80.25	77.28	76.50	79.79	62.54	77.31	77.4	74.39

Hi, there is no batch_size param in the evaluate.py, so you only use it in the commonsense evaluation, right?

Does transformers version influence the math reasoning results?
